src/serac/physics/benchmarks/benchmark_serac_thermal_functional.cpp

```diff
@@ -0,0 +1,191 @@
+  Thermal::ConstantSource source{1.0};
+  thermal_solver.setSource(source);
+
+  // Set the flux term to zero for testing code paths
+  Thermal::FluxBoundary flux_bc{0.0};
+  thermal_solver.setFluxBCs(flux_bc);
+
+  // Finalize the data structures
+  thermal_solver.completeSetup();
+
+  // Perform the quasi-static solve
+  double dt = 1.0;
+  thermal_solver.advanceTimestep(dt);
+
+  // Output the sidre-based plot files
+  thermal_solver.outputState();
+}
+
+template <int p, int dim>
+void functional_test_dynamic(double /* expected_temp_norm */)
+{
+  int serial_refinement   = 1;
+  int parallel_refinement = 0;
+
+  // Create DataStore
+  axom::sidre::DataStore datastore;
+  serac::StateManager::initialize(datastore, "thermal_functional_dynamic_solve");
+
+  static_assert(dim == 2 || dim == 3, "Dimension must be 2 or 3 for thermal functional test");
+
+  // Construct the appropriate dimension mesh and give it to the data store
+  std::string filename =
+      (dim == 2) ? SERAC_REPO_DIR "/data/meshes/star.mesh" : SERAC_REPO_DIR "/data/meshes/beam-hex.mesh";
+
+  auto mesh = mesh::refineAndDistribute(buildMeshFromFile(filename), serial_refinement, parallel_refinement);
+  serac::StateManager::setMesh(std::move(mesh));
+
+  // Define a boundary attribute set
+  std::set<int> ess_bdr = {1};
+
+  // Construct a functional-based thermal conduction solver
+  ThermalConductionFunctional<p, dim> thermal_solver(ThermalConductionFunctional<p, dim>::defaultDynamicOptions(),
+                                                     "thermal_functional");
+
+  // Define an isotropic conductor material model
+  Thermal::LinearIsotropicConductor mat(1.0, 1.0, 1.0);
+
+  thermal_solver.setMaterial(mat);
+
+  // Define the function for the initial temperature and boundary condition
+  auto initial_temp = [](const mfem::Vector& x, double) -> double {
+    if (x[0] < 0.5 || x[1] < 0.5) {
+      return 1.0;
+    }
+    return 0.0;
+  };
+
+  // Set the initial temperature and boundary condition
+  thermal_solver.setTemperatureBCs(ess_bdr, initial_temp);
+  thermal_solver.setTemperature(initial_temp);
+
+  // Define a constant source term
+  Thermal::ConstantSource source{1.0};
+  thermal_solver.setSource(source);
+
+  // Set the flux term to zero for testing code paths
+  Thermal::FluxBoundary flux_bc{0.0};
+  thermal_solver.setFluxBCs(flux_bc);
+
+  // Finalize the data structures
+  thermal_solver.completeSetup();
+
+  // Perform the time stepping
+  double dt = 0.25;
+
+  for (int i = 0; i < 4; ++i) {
+    thermal_solver.outputState();
+    thermal_solver.advanceTimestep(dt);
+  }
+
+  // Output the sidre-based plot files
+  thermal_solver.outputState();
+}
+
+}  // namespace serac
+
+//------------------------------------------------------------------------------
+
+int main(int argc, char* argv[])
+{
+  MPI_Init(&argc, &argv);
+
+  axom::slic::SimpleLogger logger;
+
+  // Initialize Adiak
+  MPI_Comm comm = MPI_COMM_WORLD;
+  adiak::init(&comm);
```

Comment:
TODO: Wrap all this adiak code in a function called something like serac::profile_init.

Reply:
Done.